File: python/python/ch3/Tree.py
import pydash

class Tree:
    def __init__(self, root):
        self._root = root
        
    @staticmethod
    def map(node, fn, tree=None):
        node.value = fn(node.value)
        if tree == None:
            tree = Tree(node)
            
        if node.has_children():
            pydash.map_(node.children, lambda child: Tree.map(child, fn, tree))
            
        return tree
      
    @property
    def root(self):
        return self._root

File: javascript/ch3/Tree.js
const _ = require('lodash');

class Tree {
  constructor(root) {
    this._root = root;
  }

  static map(node, fn, tree) {
    node.value = fn(node.value);
    if (tree == null) {
      tree = new Tree(node);
    }

    if (node.hasChildren()) {
      _.map(node.children, (child) => Tree.map(child, fn, tree));
    }

    return tree;
  }

  get root() {
    return this._root;
  }
}

module.exports = Tree;